Climber

Se-gol opened the pull request:
This branch contains code for the climber subsystem.
* get height
* set height
* get stopper mode
* get gearbox mode 
* set stopper mode
* set gearbox mode
* piston mode enum
* switch gearbox command
* switch mechanical stopper command
* pull command

Se-gol committed:
Create climber subsystem.

Se-gol committed:
Create Ports class.

Se-gol committed:
Create Constants class.

Se-gol committed:
Add getters and setters for height. Add PistonMode enum.

Add getters and setters for PistonMode.

Se-gol committed:
Add SwitchGearbox Command.

Se-gol committed:
Add SwitchMechanicalStopper Command.

Se-gol committed:
Fix Ports.

Se-gol committed:
Fix Constants.

Se-gol committed:
Add Pull Command.

orel-david commented on src/main/java/frc/robot/subsystems/climber/Climber.java:
```diff
@@ -0,0 +1,121 @@
+
+public class Climber extends SubsystemBase {
+    private final TalonFX motor = new TalonFX(Ports.Climber.MOTOR);
+    private final TalonFX salve = new TalonFX(Ports.Climber.SLAVE);
```
Did you mean salve or salve?

orel-david commented on src/main/java/frc/robot/subsystems/climber/Climber.java:
```diff
@@ -0,0 +1,121 @@
+    private final TalonFX salve = new TalonFX(Ports.Climber.SLAVE);
+    private final UnitModel unitModel = new UnitModel(Constants.Climber.TICKS_PER_METER);
+    private final Solenoid stopper = new Solenoid(Ports.Climber.STOPPER);
+    private final Solenoid gearbox = new Solenoid(Ports.Climber.GEARBOX);
```
It isn't a gearbox it is a piston

orel-david commented on src/main/java/frc/robot/subsystems/climber/Climber.java:
```diff
@@ -0,0 +1,121 @@
+import frc.robot.subsystems.UnitModel;
+
+public class Climber extends SubsystemBase {
+    private final TalonFX motor = new TalonFX(Ports.Climber.MOTOR);
```
If there is a slave this should be a master

orel-david commented on src/main/java/frc/robot/subsystems/climber/Climber.java:
```diff
@@ -0,0 +1,121 @@
+     *
+     * @param mode the stopper mode [boolean].
+     */
+    public void setStopperMode(boolean mode) {
```
I suggest that you use PistonMode here and not boolean

orel-david commented on src/main/java/frc/robot/subsystems/climber/Climber.java:
```diff
@@ -0,0 +1,121 @@
+    /**
+     * Enum for piston mode, OPEN is true, CLOSED is false.
+     */
+    public enum PistonMode {
```
Put also a toggle mode

katzuv replied:
Is toggling a mode of a piston? IMO It's an action, but not a mode.

orel-david commented on src/main/java/frc/robot/subsystems/climber/Climber.java:
```diff
@@ -0,0 +1,121 @@
+        }
+    }
+
+    @Override
+    public void periodic() {
+
+    }
```
Delete unused methods

orel-david commented on src/main/java/frc/robot/subsystems/climber/commands/Pull.java:
```diff
@@ -0,0 +1,32 @@
+
+    public Pull(Climber climber, double height) {
+        this.climber = climber;
+        addRequirements(climber);
```
Switch lines 14 and 15

orel-david commented on src/main/java/frc/robot/subsystems/climber/commands/Pull.java:
```diff
@@ -0,0 +1,32 @@
+    }
+
+    @Override
+    public void initialize() {
```
You need to disengage the stopper and make sure the gearbox is engaged

orel-david commented on src/main/java/frc/robot/subsystems/climber/commands/Pull.java:
```diff
@@ -0,0 +1,32 @@
+
+import frc.robot.subsystems.climber.Climber;
+
+public class Pull extends CommandBase {
```
Where is the isFinished?

Se-gol committed:
Fix salve to slave.

Se-gol committed:
Fix Constants.

Se-gol committed:
Fix Ports.

Se-gol committed:
Fix Climber.

Se-gol committed:
Fix SwitchGearbox.

Se-gol committed:
Fix SwitchMechanicalStopper.

Se-gol committed:
Fix Pull.

katzuv commented on src/main/java/frc/robot/Constants.java:
```diff
@@ -16,4 +16,15 @@
+        public static final int TICKS_PER_METER = 0;
+
+        public static final int kP = 0;
+
```
This newline and the one in line 25 are redundant.

katzuv commented on src/main/java/frc/robot/Constants.java:
```diff
@@ -16,4 +16,15 @@
+
+        public static final int kD = 0;
+
+        public static final double TOLERANCE = 1;
```
Be more specific.

orel-david replied:
In which units

katzuv commented on src/main/java/frc/robot/subsystems/climber/Climber.java:
```diff
@@ -0,0 +1,125 @@
+        slave.setInverted(Ports.Climber.SLAVE_INVERTED);
+
+        master.setSensorPhase(Ports.Climber.MASTER_SENSOR_PHASE_INVERTED);
+        slave.setSensorPhase(Ports.Climber.SLAVE_SENSOR_PHASE_INVERTED);
```
Does the slave have a sensor?

orel-david replied:
Yes the talonFX have an internal sensor

katzuv replied:
True, didn't think about it.
But are you using it? If not, don't configure it. You could maybe fuse the readings from the sensors, but I don't know how to feed the fused measurements back to the Talon FX, and I don't think it's necessary.

katzuv commented on src/main/java/frc/robot/subsystems/climber/Climber.java:
```diff
@@ -0,0 +1,125 @@
+     * @param height height [double].
+     */
+    public void setHeight(double height) {
+        master.set(ControlMode.Position, unitModel.toTicks(height));
```
Why not use Motion Magic?

katzuv commented on src/main/java/frc/robot/subsystems/climber/Climber.java:
```diff
@@ -0,0 +1,125 @@
+     *
+     * @return the stopper mode [PistonMode].
+     */
+    public PistonMode getStopperMode() {
```
If the function returns the "stopper mode", why is the return type `PistonMode`?
Consider renaming this method to `isStopperEngaged`and simply returning a boolean.

katzuv commented on src/main/java/frc/robot/subsystems/climber/Climber.java:
```diff
@@ -0,0 +1,125 @@
+    /**
+     * Get the climber's height.
+     *
+     * @return the climber's height [double].
```
```suggestion
     * @return the climber's height [m].
```
You don't need the `[double]`. You can include the units, but the return type is written in the [function's signature](https://en.m.wikipedia.org/wiki/Type_signature).

adamv17 replied:
"[meters]" instead of  "[double]"

katzuv commented on src/main/java/frc/robot/subsystems/climber/Climber.java:
```diff
@@ -0,0 +1,125 @@
+     *
+     * @return the gearboxPiston mode [PistonMode].
+     */
+    public PistonMode getGearboxMode() {
```
Same as above.

adamv17 commented on src/main/java/frc/robot/subsystems/climber/Climber.java:
```diff
@@ -0,0 +1,125 @@
+    private final Solenoid stopper = new Solenoid(Ports.Climber.STOPPER);
+    private final Solenoid gearboxPiston = new Solenoid(Ports.Climber.GEARBOX_PISTON);
+
+    public Climber() {
```
Config selected sensor

orel-david replied:
This are falcons they have an internal sensor

adamv17 replied:
You are right i thought they were SRX

adamv17 commented on src/main/java/frc/robot/subsystems/climber/Climber.java:
```diff
@@ -0,0 +1,125 @@
+        master.setSensorPhase(Ports.Climber.MASTER_SENSOR_PHASE_INVERTED);
+        slave.setSensorPhase(Ports.Climber.SLAVE_SENSOR_PHASE_INVERTED);
+
+        master.config_kP(0, Constants.Climber.kP);
```
Add TALON_TIMEOUT

katzuv commented on src/main/java/frc/robot/subsystems/climber/Climber.java:
```diff
@@ -0,0 +1,125 @@
+        }
+    }
+
+    public void toggleGearboxMode() {
```
```suggestion
    public void toggleGear() {
```
is enough.

katzuv commented on src/main/java/frc/robot/subsystems/climber/Climber.java:
```diff
@@ -0,0 +1,125 @@
+        gearboxPiston.set(!gearboxPiston.get());
+    }
+
+    public void toggleStopperMode() {
```
```suggestion
    public void toggleStopper() {
```
is enough.

katzuv commented on src/main/java/frc/robot/subsystems/climber/commands/SwitchGearbox.java:
```diff
@@ -0,0 +1,29 @@
+    }
+
+    @Override
+    public void execute() {
```
Do you need `execute` and `end` methods in an `InstantCommand`?

katzuv commented on src/main/java/frc/robot/subsystems/climber/commands/Pull.java:
```diff
@@ -0,0 +1,42 @@
+
+    @Override
+    public void initialize() {
+        if (climber.getStopperMode() == Climber.PistonMode.OPEN)
```
This command should only pull the climber. Another command should handle engaging the stopper, and another one will engage the shift to the relevant gear.
I recommend the two commands to be `InstanatCommand`s (as I've seen you've done), and the whole sequence (engage stopper + engage gearbox + climb) to be handled through a command group. 

This is called the _Single Responsibility Principle_, or SRP in short. In short, this means:
> The single-responsibility principle (SRP) is a computer-programming principle that states that every module, class or function in a computer program should have responsibility over a single part of that program's functionality, which it should encapsulate. All of that module, class or function's services should be narrowly aligned with that responsibility.

You can read more about the SRP [here](https://en.m.wikipedia.org/wiki/Single-responsibility_principle).

katzuv commented on src/main/java/frc/robot/subsystems/climber/commands/Pull.java:
```diff
@@ -0,0 +1,42 @@
+    private final Climber climber;
+    private double height;
+
+    public Pull(Climber climber, double height) {
```
`Climb` is a better name.

katzuv commented on src/main/java/frc/robot/subsystems/climber/commands/SwitchMechanicalStopper.java:
```diff
@@ -0,0 +1,28 @@
+import edu.wpi.first.wpilibj2.command.InstantCommand;
+import frc.robot.subsystems.climber.Climber;
+
+public class SwitchMechanicalStopper extends InstantCommand {
```
```suggestion
public class ToggleStopper extends InstantCommand {
```
Be consistent:
1) You use "toggle" anywhere else in your code.
2) You haven't mentioned the term "mechanical stopper" anywhere in your code.

barel13 replied:
don't apply the changes through github, refactor it in Intellij

katzuv replied:
For some changes that can be convenient, but in general I agree.

katzuv reviewed (changes requested):
Good job!

barel13 commented on src/main/java/frc/robot/subsystems/climber/Climber.java:
```diff
@@ -0,0 +1,125 @@
+    /**
+     * Set the climber's height.
+     *
+     * @param height height [double].
```
The purpose of JavaDoc is not to "repeat the obvious", please change the description of the height parameter.

barel13 replied:
you should also clarify what is the units of the height (meter, cm, mm, etc...)

barel13 commented on src/main/java/frc/robot/subsystems/climber/Climber.java:
```diff
@@ -0,0 +1,125 @@
+    /**
+     * Set the stopper mode.
+     *
+     * @param mode the stopper mode [boolean].
```
it's not a boolean

orel-david replied:
You don't need to specify the type because it is written below

barel13 commented on src/main/java/frc/robot/subsystems/climber/Climber.java:
```diff
@@ -0,0 +1,125 @@
+    /**
+     * Set the gearboxPiston mode.
+     *
+     * @param mode the gearboxPiston mode [boolean].
```
Again, not a boolean

barel13 commented on src/main/java/frc/robot/subsystems/climber/commands/Pull.java:
```diff
@@ -0,0 +1,42 @@
+import frc.robot.Constants;
+import frc.robot.subsystems.climber.Climber;
+
+public class Pull extends CommandBase {
```
Add description (JavaDoc) to this command

barel13 commented on src/main/java/frc/robot/subsystems/climber/commands/Pull.java:
```diff
@@ -0,0 +1,42 @@
+
+    @Override
+    public void initialize() {
+        if (climber.getStopperMode() == Climber.PistonMode.OPEN)
+            new SwitchMechanicalStopper(climber);
```
You can't call commands like that, you should use Command Group, I don't think you ever tried this, so ask @orel-david

orel-david replied:
When we want to use few commands together we use something called commandGroup. I would give you a more detailed explanation about this tomorrow, but for now  i would leave a link about this. https://docs.wpilib.org/en/stable/docs/software/commandbased/command-groups.html

barel13 commented on src/main/java/frc/robot/subsystems/climber/commands/Pull.java:
```diff
@@ -0,0 +1,42 @@
+        if (climber.getStopperMode() == Climber.PistonMode.OPEN)
+            new SwitchMechanicalStopper(climber);
+        if (climber.getGearboxMode() == Climber.PistonMode.CLOSED)
+            new SwitchGearbox(climber);
```
same as above

barel13 commented on src/main/java/frc/robot/subsystems/climber/commands/SwitchMechanicalStopper.java:
```diff
@@ -0,0 +1,28 @@
+        climber.toggleStopperMode();
+    }
+
+    @Override
+    public void execute() {
+
+    }
+
+    @Override
+    public void end(boolean interrupted) {
+
+    }
```
as @katzuv wrote before, `InstantCommand` don't need `execute` and `end` functions

barel13 commented on src/main/java/frc/robot/subsystems/climber/Climber.java:
```diff
@@ -0,0 +1,125 @@
+        master.config_kI(0, Constants.Climber.kI);
+        master.config_kD(0, Constants.Climber.kD);
+
+        slave.config_kP(0, Constants.Climber.kP);
+        slave.config_kI(0, Constants.Climber.kI);
+        slave.config_kD(0, Constants.Climber.kD);
```
The slave follows the master, you don't need to config PID for him.

barel13 reviewed (changes requested):
You write a good code. but for next time, please don't copy the name of the method in the javaDoc, it doesn't serve the purpose of documentation

orel-david commented on src/main/java/frc/robot/subsystems/climber/Climber.java:
```diff
@@ -0,0 +1,125 @@
+    }
+
+    /**
+     * Set the gearboxPiston mode.
```
```suggestion
     * Set the gearbox piston mode.

```

orel-david replied:
and give better description

adamv17 commented on src/main/java/frc/robot/subsystems/climber/Climber.java:
```diff
@@ -0,0 +1,125 @@
+
+        master.config_kP(0, Constants.Climber.kP);
+        master.config_kI(0, Constants.Climber.kI);
+        master.config_kD(0, Constants.Climber.kD);
```
Add kF to acount for gravity

katzuv replied:
Arbitrary feed forward, not kF.

orel-david commented on src/main/java/frc/robot/subsystems/climber/commands/Pull.java:
```diff
@@ -0,0 +1,42 @@
+    }
+
+    @Override
+    public void end(boolean interrupted) {
```
In the end you need to engage the stooper

Se-gol committed:
Update src/main/java/frc/robot/subsystems/climber/Climber.java

Co-authored-by: Dan Katzuv <31829093+katzuv@users.noreply.github.com>

Se-gol committed:
Update src/main/java/frc/robot/subsystems/climber/Climber.java

Co-authored-by: Dan Katzuv <31829093+katzuv@users.noreply.github.com>

Commit:
Fix Climber Class.

Add TALON_TIMEOUT, remove unnecessary lines, fix docu, fix get functions for the pistons' mode.

Commit:
Change command from set to toggle.

Commit:
Change command from set to toggle.

Commit:
Fix Climb command.

Commit:
Add AutoClimb command and insert it into Climb command.

Commit:
Add constant kF.

katzuv commented on src/main/java/frc/robot/Constants.java:
```diff
@@ -16,4 +16,17 @@
+        public static final int kD = 0;
+        public static final int kF = 0;
+
+        public static final double TOLERANCE = 1; //Error tolerance for the height requested [m].
```
```suggestion
        public static final double HIGHEST_TOLERANCE = 1; // Error tolerance for the height setpoint [m].
```
And `1m` is a pretty large tolerance…

orel-david replied:
Change the tolerance to 0.1

katzuv commented on src/main/java/frc/robot/Ports.java:
```diff
@@ -0,0 +1,23 @@
+
+    public static class Climber {
+        public static final int MASTER = 0;
+
```
Remove this newline.

katzuv commented on src/main/java/frc/robot/Ports.java:
```diff
@@ -0,0 +1,23 @@
+
+        public static final int GEARBOX_PISTON = 0;
+
+        public static final boolean MASTER_INVERTED = false;
```
```suggestion
        public static final boolean IS_MASTER_INVERTED = false;
```
Booleans may start with "is", "has", "does", etc.
The sub-team should agree on a convention (@orel-david).

katzuv commented on src/main/java/frc/robot/Ports.java:
```diff
@@ -0,0 +1,23 @@
+
+        public static final int STOPPER = 0;
+
+        public static final int GEARBOX_PISTON = 0;
```
Ask the other guys, but I think `GEARBOX_SHIFTER` is a better name. "piston" is an implementation detail which is not interesting. 
@orel-david, what do you think?

katzuv commented on src/main/java/frc/robot/Ports.java:
```diff
@@ -0,0 +1,23 @@
+
+        public static final boolean SLAVE_INVERTED = false;
+
+        public static final boolean SLAVE_SENSOR_PHASE_INVERTED = false;
```
Do you use the slave's sensor?

katzuv commented on src/main/java/frc/robot/subsystems/climber/Climber.java:
```diff
@@ -0,0 +1,117 @@
+    private final TalonFX slave = new TalonFX(Ports.Climber.SLAVE);
+    private final UnitModel unitModel = new UnitModel(Constants.Climber.TICKS_PER_METER);
+    private final Solenoid stopper = new Solenoid(Ports.Climber.STOPPER);
+    private final Solenoid gearboxPiston = new Solenoid(Ports.Climber.GEARBOX_PISTON);
```
A couple of things:
1) `gearbox` or `gearShifter`, like you didn't write `stopperPiston`. It's an unintersecting implementation detail.
2) This gearbox and its motors are used by the shooter, too. I think it should be in a separate subsystem. You need to think how you'll make this subsystem (the gearbox) communicate effectively with both the shooter and the climber, maybe by using [suppliers](https://docs.oracle.com/javase/8/docs/api/java/util/function/Supplier.html) and lambda functions. 
I'm not sure that's the best solution for this complicated situation (motors used by multiple subsystems). Consult your teammates and @orel-david.

katzuv commented on src/main/java/frc/robot/subsystems/climber/Climber.java:
```diff
@@ -0,0 +1,117 @@
+     * @param height requested height to climb [m].
+     */
+    public void setHeight(double height) {
+        master.set(ControlMode.MotionMagic, unitModel.toTicks(height));
```
You need to configure the _arbitrary feed forward_ value. If you aren't familiar with it, you can read more [here](https://phoenix-documentation.readthedocs.io/en/latest/ch16_ClosedLoop.html#arbitrary-feed-forward) and ask your teammates.

katzuv replied:
You need to configure the _arbitrary feed forward_ value. If you aren't familiar with it, you can read more [here](https://phoenix-documentation.readthedocs.io/en/latest/ch16_ClosedLoop.html#arbitrary-feed-forward) and ask your teammates.

orel-david replied:
For motion magic you also need to configure acceleration and cruise velocity in the constructor

katzuv commented on src/main/java/frc/robot/subsystems/climber/Climber.java:
```diff
@@ -0,0 +1,117 @@
+    }
+
+    /**
+     * Get whether or not the stopper is engaged.
```
```suggestion
     * Get whether the stopper is engaged.
```
is enough.

katzuv commented on src/main/java/frc/robot/subsystems/climber/Climber.java:
```diff
@@ -0,0 +1,117 @@
+    /**
+     * Get whether or not the stopper is engaged.
+     *
+     * @return whether or not the stopper is engaged.
```
Don't just duplicate the documentation. I think leaving just the `@return` part is enough.

katzuv commented on src/main/java/frc/robot/subsystems/climber/Climber.java:
```diff
@@ -0,0 +1,117 @@
+    }
+
+    /**
+     * Get whether or not the gearbox piston is engaged.
+     *
+     * @return whether or not the gearbox piston is engaged.
```
Upper two comments.

katzuv commented on src/main/java/frc/robot/subsystems/climber/Climber.java:
```diff
@@ -0,0 +1,117 @@
+     *
+     * @param mode the stopper mode.
+     */
+    public void setStopperMode(PistonMode mode) {
+        stopper.set(mode.getValue());
+    }
+
+    /**
+     * Set the gearbox piston mode to a given mode.
+     *
+     * @param mode the wanted gearbox piston mode.
+     */
+    public void setGearboxMode(PistonMode mode) {
```
If the getters return booleans, then the setters should receive booleans too.

katzuv commented on src/main/java/frc/robot/subsystems/climber/commands/SetGearbox.java:
```diff
@@ -0,0 +1,20 @@
+    private final Climber climber;
+    private Climber.PistonMode mode;
+
+    public SetGearbox(Climber climber, Climber.PistonMode mode) {
```
1) I think `ShiftGear` is a better and more concise name. 
2) Be consistent. You can't call the the command `SetGearbox`, but supply it a `PistonMode`. Isn't a boolean enough?

katzuv commented on src/main/java/frc/robot/subsystems/climber/commands/Climb.java:
```diff
@@ -0,0 +1,42 @@
+
+    @Override
+    public void initialize() {
+        new AutoClimb(climber);
```
Use a command group for the whole sequence instead. 
I also don't think it'd work because both `AutoClimb` and `Climb` require the climber.

orel-david replied:
You should add `climb` to `autoClimb` and not the opposite.

katzuv commented on src/main/java/frc/robot/subsystems/climber/commands/AutoClimb.java:
```diff
@@ -0,0 +1,14 @@
+
+    private final Climber climber;
+
+    public AutoClimb(Climber climber) {
```
1) You have both `Climb` and `AutoClimb`. The commands' names don't explain their differences.
2) This command name contains the word "climb", but there is no actual climbing that runs in it. It only shifts the gear and disengages the stopper.

orel-david commented on src/main/java/frc/robot/subsystems/climber/Climber.java:
```diff
@@ -0,0 +1,117 @@
+
+        master.config_kP(0, Constants.Climber.kP, Constants.TALON_TIMEOUT);
+        master.config_kI(0, Constants.Climber.kI, Constants.TALON_TIMEOUT);
+        master.config_kD(0, Constants.Climber.kD, Constants.TALON_TIMEOUT);
```
Add Kf for the master

Commit:
Fix TOLERANCE constant.

Commit:
Change naming scheme.

Se-gol committed:
Update src/main/java/frc/robot/subsystems/climber/Climber.java

Co-authored-by: Dan Katzuv <31829093+katzuv@users.noreply.github.com>

Commit:
Add CRUISE_VELOCITY & ACCELERATION constants.

Commit:
Change name to ShiftGear.

Commit:
Change command to manage climb.

Commit:
Fix auto climb command.

Commit:
Fix docu and configs.

Commit:
Remove unnecessary line.

Commit:
Merge remote-tracking branch 'origin/climber' into climber

barel13 commented on src/main/java/frc/robot/Constants.java:
```diff
@@ -16,4 +16,20 @@
  */
 public final class Constants {
+
+    public static final int TALON_TIMEOUT = 10; //Waiting period for configurations [ns].
```
it should be in milliseconds not nanoseconds

barel13 commented on src/main/java/frc/robot/Constants.java:
```diff
@@ -16,4 +16,20 @@
+        public static final int kF = 0;
+
+        public static final int CRUISE_VELOCITY = 0;
+        public static final int ACCELERATION = 0;
```
add units

barel13 commented on src/main/java/frc/robot/Ports.java:
```diff
@@ -0,0 +1,16 @@
+        public static final boolean IS_MASTER_INVERTED = false;
+        public static final boolean IS_MASTER_SENSOR_PHASE_INVERTED = false;
+        public static final boolean IS_SLAVE_INVERTED = false;
+//        public static final boolean IS_SLAVE_SENSOR_PHASE_INVERTED = false;
```
if you don't use it, delete it

barel13 commented on src/main/java/frc/robot/subsystems/climber/Climber.java:
```diff
@@ -0,0 +1,122 @@
+        slave.setInverted(Ports.Climber.IS_SLAVE_INVERTED);
+
+        master.setSensorPhase(Ports.Climber.IS_MASTER_SENSOR_PHASE_INVERTED);
+//        slave.setSensorPhase(Ports.Climber.IS_SLAVE_SENSOR_PHASE_INVERTED);
```
same as above, just delete it

barel13 commented on src/main/java/frc/robot/subsystems/climber/Climber.java:
```diff
@@ -0,0 +1,122 @@
+    }
+
+    /**
+     * ManageClimb height meters up with the robot.
```
don't write 2 words as one in the documentation

barel13 commented on src/main/java/frc/robot/subsystems/climber/Climber.java:
```diff
@@ -0,0 +1,122 @@
+    }
+
+    /**
+     * Get whether the stopper is engaged.
```
```suggestion
     * Whether the stopper is engaged.

```

katzuv replied:
The "get" is important. That's what the method does.

barel13 commented on src/main/java/frc/robot/subsystems/climber/Climber.java:
```diff
@@ -0,0 +1,122 @@
+    }
+
+    /**
+     * Get whether the gearbox shifter is engaged.
```
same as above

barel13 commented on src/main/java/frc/robot/subsystems/climber/Climber.java:
```diff
@@ -0,0 +1,122 @@
+        }
+    }
+
+    public void toggleGear() {
+        gearboxShifter.set(!gearboxShifter.get());
+    }
+
+    public void toggleStopper() {
+        stopper.set(!stopper.get());
+    }
```
Add documentation

barel13 commented on src/main/java/frc/robot/subsystems/climber/commands/AutoClimb.java:
```diff
@@ -0,0 +1,14 @@
+
+public class AutoClimb extends SequentialCommandGroup {
+
+    private final Climber climber;
```
If you don't use this object, why did you instantiated it outside of the constructor

barel13 commented on src/main/java/frc/robot/subsystems/climber/commands/ManageClimb.java:
```diff
@@ -0,0 +1,41 @@
+
+    @Override
+    public void end(boolean interrupted) {
+        new SetStopper(climber, Climber.PistonMode.CLOSED);
```
don't create commands inside, instead add it to the command group

Commit:
Fix colour_wheel.

Commit:
Fix climber.

Commit:
Merge remote-tracking branch 'origin/climber' into climber

barel13 committed:
Fix documentation

p.s it's Saar

orel-david commented on src/main/java/frc/robot/Constants.java:
```diff
@@ -16,4 +16,20 @@
+        public static final int kF = 0;
+
+        public static final int CRUISE_VELOCITY = 0; //sensor units per 100ms.
+        public static final int ACCELERATION = 0; //sensor units per 100ms.
```
```suggestion
        public static final int ACCELERATION = 0; //sensor units per 100ms^2.
```

orel-david commented on src/main/java/frc/robot/Constants.java:
```diff
@@ -16,4 +16,20 @@
+    public static class Climber {
+        public static final int TICKS_PER_METER = 0;
+
+        public static final int kP = 0;
+        public static final int kI = 0;
+        public static final int kD = 0;
+        public static final int kF = 0;
```
Constants should be all capital

orel-david commented on src/main/java/frc/robot/Ports.java:
```diff
@@ -0,0 +1,15 @@
+        public static final int GEARBOX_SHIFTER = 0;
+
+        public static final boolean IS_MASTER_INVERTED = false;
+        public static final boolean IS_MASTER_SENSOR_PHASE_INVERTED = false;
```
Only the master has sensor so you can change it to `SENSOR_PHASE`

orel-david commented on src/main/java/frc/robot/subsystems/climber/commands/AutoClimb.java:
```diff
@@ -0,0 +1,11 @@
+public class AutoClimb extends SequentialCommandGroup {
+
+    public AutoClimb(Climber climber, double height) {
+        addCommands(new ShiftGear(climber, Climber.PistonMode.OPEN), new SetStopper(climber, Climber.PistonMode.CLOSED), new ManageClimb(climber, height), new SetStopper(climber, Climber.PistonMode.CLOSED));
```
```suggestion
        addCommands(new ShiftGear(climber, Climber.PistonMode.OPEN), 
        new SetStopper(climber, Climber.PistonMode.CLOSED),
         new ManageClimb(climber, height),
          new SetStopper(climber, Climber.PistonMode.CLOSED));

```

orel-david replied:
And in line 10 you should disengage the stopper so you can move.

Se-gol committed:
Naming fixes and auto climb fix.

barel13 commented on src/main/java/frc/robot/subsystems/climber/Climber.java:
```diff
@@ -0,0 +1,124 @@
+     * Enum for piston mode, OPEN is true, CLOSED is false.
+     */
+    public enum PistonMode {
+        OPEN(true), CLOSED(false);
```
Change the `CLOSED` state to `CLOSE`. You need to follow the same conventions for both states (i.e both of them need to be in the same time clause)

Se-gol replied:
![716Y5Cw1L3L _SL1500_](https://user-images.githubusercontent.com/71329836/103660685-90dbef00-4f76-11eb-9f55-ca113e715ede.jpg)

barel13 commented on src/main/java/frc/robot/subsystems/climber/Climber.java:
```diff
@@ -0,0 +1,124 @@
+         */
+        public boolean getValue() {
+            return on;
+        }
```
It's not necessary, but you can make the boolean on to final object, and then make him private instead of using this function, because boolean is an immutable object (You can't change his value, in this case because he may declared `final`)

barel13 commented on src/main/java/frc/robot/subsystems/climber/commands/SetStopper.java:
```diff
@@ -0,0 +1,20 @@
+import edu.wpi.first.wpilibj2.command.InstantCommand;
+import frc.robot.subsystems.climber.Climber;
+
+public class SetStopper extends InstantCommand {
```
add documentation

barel13 commented on src/main/java/frc/robot/subsystems/climber/commands/ShiftGear.java:
```diff
@@ -0,0 +1,20 @@
+import edu.wpi.first.wpilibj2.command.InstantCommand;
+import frc.robot.subsystems.climber.Climber;
+
+public class ShiftGear extends InstantCommand {
```
add documentation

barel13 commented on src/main/java/frc/robot/Constants.java:
```diff
@@ -16,4 +16,20 @@
+    public static class Climber {
+        public static final int TICKS_PER_METER = 0;
+
+        public static final int KP = 0;
+        public static final int KI = 0;
+        public static final int KD = 0;
+        public static final int KF = 0;
+
+        public static final int CRUISE_VELOCITY = 0; //sensor units per 100ms.
+        public static final int ACCELERATION = 0; //sensor units per 100ms^2.
```
should be double (everything I commented)

barel13 commented on src/main/java/frc/robot/subsystems/climber/Climber.java:
```diff
@@ -0,0 +1,124 @@
+    /**
+     * Enum for piston mode, OPEN is true, CLOSED is false.
+     */
+    public enum PistonMode {
```
For 2 modes you don't need an enum, you can use a boolean

orel-david replied:
Actually it makes it clearer so i like it.

barel13 commented on src/main/java/frc/robot/subsystems/climber/commands/ManageClimb.java:
```diff
@@ -0,0 +1,40 @@
+        climber.setHeight(height);
+    }
+
+    @Override
+    public void execute() {
+
+    }
```
delete empty method

barel13 commented on src/main/java/frc/robot/subsystems/climber/commands/ManageClimb.java:
```diff
@@ -0,0 +1,40 @@
+        return Math.abs(climber.getHeight() - height) <= Constants.Climber.TOLERANCE;
+    }
+
+    @Override
+    public void end(boolean interrupted) {
+    }
```
delete empty method

orel-david commented on src/main/java/frc/robot/subsystems/climber/Climber.java:
```diff
@@ -0,0 +1,124 @@
+    private final TalonFX slave = new TalonFX(Ports.Climber.SLAVE);
+    private final UnitModel unitModel = new UnitModel(Constants.Climber.TICKS_PER_METER);
+    private final Solenoid stopper = new Solenoid(Ports.Climber.STOPPER);
+    private final Solenoid gearboxShifter = new Solenoid(Ports.Climber.GEARBOX_SHIFTER);
```
The gearbox shifter wouldn't be part of the climber it would be part of the PTO

Se-gol replied:
?

orel-david commented on src/main/java/frc/robot/subsystems/climber/Climber.java:
```diff
@@ -0,0 +1,124 @@
+    private final TalonFX master = new TalonFX(Ports.Climber.MASTER);
+    private final TalonFX slave = new TalonFX(Ports.Climber.SLAVE);
+    private final UnitModel unitModel = new UnitModel(Constants.Climber.TICKS_PER_METER);
+    private final Solenoid stopper = new Solenoid(Ports.Climber.STOPPER);
```
Today i talked to liron and we are switching to double solenoids instead of solenoids

orel-david commented on src/main/java/frc/robot/subsystems/climber/commands/AutoClimb.java:
```diff
@@ -0,0 +1,14 @@
+
+    public AutoClimb(Climber climber, double height) {
+        addCommands(new ShiftGear(climber, Climber.PistonMode.OPEN),
+                new SetStopper(climber, Climber.PistonMode.OPEN),
+                new ManageClimb(climber, height),
+                new SetStopper(climber, Climber.PistonMode.CLOSED));
```
You can add the set stopper into the manage climb command

barel13 commented on src/main/java/frc/robot/Constants.java:
```diff
@@ -16,4 +16,20 @@
+        public static final int CRUISE_VELOCITY = 0; //sensor units per 100ms.
+        public static final int ACCELERATION = 0; //sensor units per 100ms^2.
+
+        public static final double TOLERANCE = 0.1; //Error tolerance for the height [m].
```
When using this variable (not reading the comment) it's hard to know the purpose of the TOLERANCE. Is it tolerance of the height, of the velocity, the encoder, etc...

barel13 commented on src/main/java/frc/robot/subsystems/climber/Climber.java:
```diff
@@ -0,0 +1,124 @@
+     *
+     * @return whether the stopper is engaged.
+     */
+    public boolean isStoppedEngaged() {
+        return stopper.get();
+    }
+
+    /**
+     * Get whether the gearbox shifter is engaged.
+     *
+     * @return whether the gearbox shifter is engaged.
+     */
+    public boolean isGearboxEngaged() {
+        return gearboxShifter.get();
+    }
```
If you are using the `PistonMode`, us it everywhere or nowhere.

barel13 commented on src/main/java/frc/robot/subsystems/climber/commands/ManageClimb.java:
```diff
@@ -0,0 +1,31 @@
+public class ManageClimb extends CommandBase {
+
+    private final Climber climber;
+    private double height;
```
```suggestion
    private final double height;

```

Se-gol committed:
Fixing Climber

orel-david commented on src/main/java/frc/robot/subsystems/climber/commands/AutoClimb.java:
```diff
@@ -0,0 +1,13 @@
+    public AutoClimb(Climber climber, double height) {
+        addCommands(new ShiftGear(climber, Climber.PistonMode.OPEN),
+                new SetStopper(climber, Climber.PistonMode.OPEN),
+                new ManageClimb(climber, height));
```
You need to lock the climber at the end

orel-david commented on src/main/java/frc/robot/subsystems/climber/commands/ManageClimb.java:
```diff
@@ -0,0 +1,41 @@
+
+    @Override
+    public void initialize() {
+        new SetStopper(climber, Climber.PistonMode.CLOSED);
```
Don't call another command in the initialize and use command group

barel13 commented on src/main/java/frc/robot/Constants.java:
```diff
@@ -16,4 +16,20 @@
+    public static class Climber {
+        public static final int TICKS_PER_METER = 0;
+
+        public static final int KP = 0;
+        public static final int KI = 0;
+        public static final int KD = 0;
+        public static final int KF = 0;
```
double it

barel13 commented on src/main/java/frc/robot/subsystems/climber/Climber.java:
```diff
@@ -0,0 +1,119 @@
+    private final TalonFX master = new TalonFX(Ports.Climber.MASTER);
+    private final TalonFX slave = new TalonFX(Ports.Climber.SLAVE);
+    private final UnitModel unitModel = new UnitModel(Constants.Climber.TICKS_PER_METER);
+    private final DoubleSolenoid stopper = new DoubleSolenoid(Ports.Climber.STOPPER, Ports.Climber.STOPPER_2);
```
STOPPER_2 ?????

barel13 commented on src/main/java/frc/robot/subsystems/climber/Climber.java:
```diff
@@ -0,0 +1,119 @@
+        if (mode.getValue())
+            stopper.set(DoubleSolenoid.Value.kForward);
+        else
+            stopper.set(DoubleSolenoid.Value.kOff);
```
IIRC you need kReverse

barel13 commented on src/main/java/frc/robot/subsystems/climber/Climber.java:
```diff
@@ -0,0 +1,119 @@
+     */
+    public enum PistonMode {
+        OPEN(true), CLOSED(false);
+        private boolean on;
```
```suggestion
        private final boolean on;

```

barel13 commented on src/main/java/frc/robot/subsystems/climber/commands/SetStopper.java:
```diff
@@ -0,0 +1,20 @@
+
+public class SetStopper extends InstantCommand {
+    private final Climber climber;
+    private Climber.PistonMode mode;
```
```suggestion
    private final Climber.PistonMode mode;

```

barel13 commented on src/main/java/frc/robot/subsystems/climber/commands/ShiftGear.java:
```diff
@@ -0,0 +1,20 @@
+
+public class ShiftGear extends InstantCommand {
+    private final Climber climber;
+    private Climber.PistonMode mode;
```
```suggestion
    private final Climber.PistonMode mode;

```

Se-gol committed:
General Fixes to Climber.

orel-david commented on src/main/java/frc/robot/subsystems/climber/commands/AutoClimb.java:
```diff
@@ -0,0 +1,13 @@
+    public AutoClimb(Climber climber, double height) {
+        addCommands(new ShiftGear(climber, Climber.PistonMode.OPEN),
+                new SetStopper(climber, Climber.PistonMode.OPEN),
+                new ManageClimb(climber, height), new SetStopper(climber, Climber.PistonMode.CLOSED));
```
```suggestion
                new ManageClimb(climber, height),
                 new SetStopper(climber, Climber.PistonMode.CLOSED));

```

Se-gol committed:
Update src/main/java/frc/robot/subsystems/climber/commands/AutoClimb.java

Co-authored-by: orel-david <42713127+orel-david@users.noreply.github.com>

barel13 commented on src/main/java/frc/robot/subsystems/climber/Climber.java:
```diff
@@ -0,0 +1,119 @@
+import frc.robot.Ports;
+import frc.robot.subsystems.UnitModel;
+
+public class Climber extends SubsystemBase {
```
Add documentation

barel13 commented on src/main/java/frc/robot/subsystems/climber/Climber.java:
```diff
@@ -0,0 +1,119 @@
+    }
+
+    /**
+     * climb height meters up with the robot.
```
```suggestion
     * Move the climber to a given height.
```

barel13 commented on src/main/java/frc/robot/subsystems/climber/Climber.java:
```diff
@@ -0,0 +1,119 @@
+         *
+         * @return the value of the piston mode [boolean].
+         */
+        public boolean getValue() {
```
Rename the function, it's hard to understand what is the "value", how to use it and why it's important

barel13 commented on src/main/java/frc/robot/subsystems/climber/Climber.java:
```diff
@@ -0,0 +1,119 @@
+        master.set(ControlMode.MotionMagic, unitModel.toTicks(height));
+    }
+
+    /**
+     * Set the stopper mode.
+     *
+     * @param mode the stopper mode.
+     */
```
```suggestion
    /**
     * Set the stopper shifter mode to a given mode.
     *
     * @param mode the wanted stopper shifter mode.
     */
```

# **Be consistent!**

barel13 commented on src/main/java/frc/robot/subsystems/climber/Climber.java:
```diff
@@ -0,0 +1,119 @@
+            gearboxShifter.set(DoubleSolenoid.Value.kForward);
+        else
+            gearboxShifter.set(DoubleSolenoid.Value.kReverse);
+
```
Remove empty line

barel13 commented on src/main/java/frc/robot/subsystems/climber/Climber.java:
```diff
@@ -0,0 +1,119 @@
+    }
+
+    /**
+     * Enum for piston mode, OPEN is true, CLOSED is false.
```
```suggestion
     * Enum for piston mode. Stores whether the piston is open or not via the "on" boolean.

```

barel13 commented on src/main/java/frc/robot/subsystems/climber/Climber.java:
```diff
@@ -0,0 +1,119 @@
+    /**
+     * Enum for piston mode, OPEN is true, CLOSED is false.
+     */
+    public enum PistonMode {
```
Move the enum to the end of the file

barel13 commented on src/main/java/frc/robot/subsystems/climber/commands/ShiftGear.java:
```diff
@@ -0,0 +1,23 @@
+/**
+ * Set the gearbox to be open or closed.
+ */
+public class ShiftGear extends InstantCommand {
```
The PTO should handle the gear, not the climber!

barel13 commented on src/main/java/frc/robot/subsystems/climber/Climber.java:
```diff
@@ -0,0 +1,119 @@
+    private final TalonFX slave = new TalonFX(Ports.Climber.SLAVE);
+    private final UnitModel unitModel = new UnitModel(Constants.Climber.TICKS_PER_METER);
+    private final DoubleSolenoid stopper = new DoubleSolenoid(Ports.Climber.STOPPER_FORWARD_CHANNEL, Ports.Climber.STOPPER_REVERSE_CHANNEL);
+    private final DoubleSolenoid gearboxShifter = new DoubleSolenoid(Ports.Climber.GEARBOX_FORWARD_CHANNEL, Ports.Climber.GEARBOX_REVERSE_CHANNEL);
```
You don't need it, the PTO handles the gearbox

Se-gol committed:
More fixes to climber.

barel13 commented on src/main/java/frc/robot/subsystems/climber/Climber.java:
```diff
@@ -0,0 +1,121 @@
+     */
+    public enum PistonMode {
+        OPEN(true), CLOSED(false);
+        private final boolean on;
```
Why do you need this boolean? you need it (you can just compare to the enum)

barel13 commented on src/main/java/frc/robot/subsystems/climber/commands/ManageClimb.java:
```diff
@@ -0,0 +1,31 @@
+import frc.robot.subsystems.climber.Climber;
+
+/**
+ * The command lifts the robot up by a given number of meters.
```
"by a given number of meters" or "to a certain height"?

Se-gol committed:
Remove unnecessary variable from PistonMode enum.

barel13 reviewed (approved):
Just fix the documentation.

adamv17 commented on src/main/java/frc/robot/subsystems/climber/commands/AutoClimb.java:
```diff
@@ -0,0 +1,15 @@
+        addCommands(new ShiftGear(climber, Climber.PistonMode.OPEN),
+                new SetStopper(climber, Climber.PistonMode.OPEN),
+                new ManageClimb(climber, height),
+                 new SetStopper(climber, Climber.PistonMode.CLOSED));
```
format

Se-gol committed:
Update src/main/java/frc/robot/subsystems/climber/Climber.java

Co-authored-by: Barel Shlidor <42340005+barel13@users.noreply.github.com>

Se-gol committed:
Final fixes to Climber.

Se-gol committed:
Merge remote-tracking branch 'origin/climber' into climber
